lib/sdk/domain/biocentral_database.dart

```diff
@@ -52,6 +56,80 @@ abstract class BiocentralDatabase<T extends BioEntity> {
     return result;
   }
 
+  bool isNumeric(Map<String, dynamic> columnValues) {
+    final nonNullValues = columnValues.values
+        .where((value) => value != null && value.toString() != 'Unknown')
+        .toList();
+
+    if (nonNullValues.isEmpty) return false;
+
+    // To prevent the case where 0/1 only is tagged as numeric column
+    return !isBinary(columnValues) &&
+        nonNullValues.every((value) {
+          final String strValue = value.toString().trim();
+          return num.tryParse(strValue) != null;
+        });
+  }
+
+  bool isBinary(Map<String, dynamic> columnValues) {
+    final nonNullValues = columnValues.values
+        .where((value) => value != null && value.toString() != 'Unknown')
+        .map((value) => value.toString().trim())
+        .toSet();
+
+    return nonNullValues.length == 2;
+  }
+
+  /// Get the trainable column names, optionally filtered by type
+  /// A column is trainable if it has at least one null or "Unknown" value
+  ///
+  /// Parameters:
+  /// - [binaryTypes]: If true, only include binary columns
+  /// - [numericTypes]: If true, only include numeric columns
+  /// Add more types here as needed
+  List<String> getTrainableColumnNames(
```

Comment:
```suggestion
  List<String> getPartiallyUnlabeledColumnNames(
```
I think that "trainable" is a bit misleading regarding the wording, because for actual supervised training, not BO, we do need columns that don't have missing values.